core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -265,6 +266,52 @@ public void testPrimaryPhaseExecutesRequest() throws InterruptedException, Execu
+        clusterService.setState(state(index, true, ShardRoutingState.RELOCATING));
+        Request request = new Request(shardId).timeout("1ms");
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        TransportReplicationAction.PrimaryPhase primaryPhase = action.new PrimaryPhase(request, createTransportChannel(listener));
+        isRelocated.set(true);
+        primaryPhase.run();
+        assertThat("request was processed on primary", request.processedOnPrimary.get(), equalTo(false));
+        final String relocatingNodeId = clusterService.state().getRoutingTable().shardRoutingTable(index, shardId.id()).primaryShard().relocatingNodeId();
+        final List<CapturingTransport.CapturedRequest> requests = transport.capturedRequestsByTargetNode().get(relocatingNodeId);
+        assertThat(requests, notNullValue());
+        assertThat(requests.size(), equalTo(1));
+        assertThat("primary request was not sent", requests.get(0).action, equalTo("testAction[p]"));
+    }
+
+    public void testPrimaryPhaseExecutesDelegatedRequest() throws InterruptedException, ExecutionException {
+        final String index = "test";
+        final ShardId shardId = new ShardId(index, 0);
+        ClusterState state = state(index, true, ShardRoutingState.RELOCATING, ShardRoutingState.STARTED);
+        String primarySourceNodeId = state.getRoutingTable().shardRoutingTable(shardId.getIndex(), shardId.id()).primaryShard().currentNodeId();
+        String primaryTargetNodeId = state.getRoutingTable().shardRoutingTable(shardId.getIndex(), shardId.id()).primaryShard().relocatingNodeId();
+        String replicaNodeId = state.getRoutingTable().shardRoutingTable(index, shardId.id()).replicaShards().get(0).currentNodeId();
+        // simulate execution of the primary phase on the relocation target node
+        state = ClusterState.builder(state).nodes(DiscoveryNodes.builder(state.nodes()).localNodeId(primaryTargetNodeId)).build();
+        clusterService.setState(state);
+        Request request = new Request(shardId).timeout("1ms");
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        TransportReplicationAction.PrimaryPhase primaryPhase = action.new PrimaryPhase(request, createTransportChannel(listener));
+        primaryPhase.run();
+        assertThat("request was not processed on primary", request.processedOnPrimary.get(), equalTo(true));
+
+        // check that normal replication to replicas works
+        List<CapturingTransport.CapturedRequest> requests = transport.capturedRequestsByTargetNode().get(replicaNodeId);
+        assertThat(requests, notNullValue());
+        assertThat(requests.size(), equalTo(1));
+        assertThat("replica request was not sent to replica", requests.get(0).action, equalTo("testAction[r]"));
+
+        // check that we also replicate back to primary relocation source
+        requests = transport.capturedRequestsByTargetNode().get(primarySourceNodeId);
```

Comment:
this should be part of the tests for the replication phase - with random cluster state and all..